compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ClassCodegen.kt

```diff
@@ -216,7 +228,55 @@ abstract class ClassCodegen protected constructor(
         return null
     }
 
-    protected abstract fun generateKotlinMetadataAnnotation()
+    private val metadataSerializer: MetadataSerializer =
+        context.serializerFactory(context, irClass, type, visitor.serializationBindings, parentClassCodegen?.metadataSerializer)
+
+    private fun generateKotlinMetadataAnnotation() {
+        var extraFlags = JvmAnnotationNames.METADATA_JVM_IR_FLAG
+        if (state.isIrWithStableAbi) {
+            extraFlags += JvmAnnotationNames.METADATA_JVM_IR_STABLE_ABI_FLAG
+        }
+
+        val facadeClassName = context.multifileFacadeForPart[irClass.attributeOwnerId]
+        val metadata = irClass.metadata
+        val entry = irClass.fileParent.fileEntry
+        val kind = when {
+            facadeClassName != null -> KotlinClassHeader.Kind.MULTIFILE_CLASS_PART
+            metadata is MetadataSource.Class -> KotlinClassHeader.Kind.CLASS
+            metadata is MetadataSource.File -> KotlinClassHeader.Kind.FILE_FACADE
+            metadata is MetadataSource.Function -> KotlinClassHeader.Kind.SYNTHETIC_CLASS
+            entry is MultifileFacadeFileEntry -> KotlinClassHeader.Kind.MULTIFILE_CLASS
+            else -> KotlinClassHeader.Kind.SYNTHETIC_CLASS
+        }
+        writeKotlinMetadata(visitor, state, kind, extraFlags) {
```

Comment:
Note that there's a minor change in behavior in that we'll write the `METADATA_JVM_IR_FLAG`/`METADATA_JVM_IR_STABLE_ABI_FLAG` to synthetic classes too (previously we used `writeSyntheticClassMetadata` which passed extraFlags=0).

But I think the new behavior is more useful, so let's keep it. It doesn't affect anything anyway, other than manual bytecode introspection.

Reply:
Yes, I noticed the 0 but thought that it looked like an oversight.